### DISCUSSION TITLE
Automatic Differentiation of object oriented paradigms

### DISCUSSION THREAD

**Discussion by kaushal-malpure** [vgvassilev/clad#804] [Q&A]
Hey @parth-07 @vgvassilev  ,
I wanted to know more about the differentiation of object oriented paradigms and how exactly clad works on it.
Moreover how we approach on setting up support for differentiation of an object oriented paradigm. I would be grateful if you could help me with this.

**Comment by parth-07** [comment]
Hi @kaushal-malpure 

Differentiation of object-oriented constructs simply means that Clad is able to generated derivative code of functions that uses object-oriented features. Object-oriented feature includes: classes, inheritance, virtual functions, member functions, operator overloads, special member functions, and so on. Clad already supports some of these constructs: basic classes, member functions, operator overloads and virtual functions to some extent. 

> how exactly clad works on it.
> Moreover how we approach on setting up support for differentiation of an object oriented paradigm

I will suggest you to ask more targeted questions. These questions seems to be too open-ended to be useful. 

Please refer to Clad tests to see how Clad differentiates codes containing class features.

**Reply by kaushal-malpure** [reply]
Hi @parth-07 @vgvassilev ,
I tried understanding the test for member functions in forward mode(`test/ForwardMode/MemberFunctions.C`) and reverse mode(`test/Gradient/MemberFunctions.C`) and understood most of the part. But I didn't get how the clad::differentiate or clad::gradient works for the differentiation of member functions? 
```
SimpleFunctions sf(2, 3);
  auto d_addValueToSimpleFunction = clad::differentiate(addValueToSimpleFunction, "value");
  auto d_subtractValueFromSimpleFunction = clad::differentiate(subtractValueFromSimpleFunction, "value");
  auto d_multiplySimpleFunctionByValue = clad::differentiate(multiplySimpleFunctionByValue, "value");
```


Can you please suggest me any readings or file where I can understand its functioning.
